torch_geometric/nn/models/basic_gnn.py

```diff
@@ -118,11 +118,16 @@ def reset_parameters(self):
         if hasattr(self, 'lin'):
             self.lin.reset_parameters()
 
-    def forward(self, x: Tensor, edge_index: Adj, *args, **kwargs) -> Tensor:
+    def forward(self, x: Tensor, edge_index: Adj, *args,
```

Comment:
I'm wondering if we solely should add this feature to `GAT` rather than `BasicGNN`. Alternatively, we could make `GAT` inherit from `BasicAttentionalGNN` that again inherits from `BasicGNN`. IMO, I'm not that comfortable adding this feature to every GNN.

Alternatively, on could add a post forward hook to all `GATConv` operators and aggregate the results there. This does not require modifications to `forward`.

It might be also a good idea to document this feature inside the doc-string of `forward`.

Reply:
I understand the issue regarding implementing GAT-specific components in `BasicGNN`. Creating a `BasicAttentionalGNN` seems to be an overkill and not sure how to use hooks here, so what I did is to modify the forward method of `GATConv`, which now only returns `out`. It also has now a getter to return attention weights based on the latest `forward` call. 

Not sure if this is the best way to do it, let me know what you think! Also, we could then get rid of the `return_attention_weights` argument in the `forward` function. I also don't understand the note regarding this argument: https://github.com/pyg-team/pytorch_geometric/blob/d1aa1837b36dceed9349ec083bba807ac3f58f3f/torch_geometric/nn/conv/gat_conv.py#L189

Reply:
P.S. `test_gat_conv` fails for now, I'll modify it if you think this is the good way to go

Reply:
I like the idea of just adding this to `GAT` personally, it seems like a very specific use case (actually I think it would be even okay to simple have)
```
def forward(x, **kwargs):
     if "return_attention_weights" in kawargs:
           raise NotImplementedException("...")
    super().forward(x, **kwargs)
```

inside of the GAT class.

Reply:
I agree. I think the best solution would be to leverage `forward` hooks on each `GATConv` module in `GAT` which collects all attention weights of all layers, and additionally returns them in case `return_attention_weights=True`.